packages/next-on-pages/templates/_worker.js/utils/cache-interface.ts

```diff
@@ -0,0 +1,248 @@
+
+			if (revalidatedAt) {
+				data.revalidatedAt = revalidatedAt;
+			}
+
+			tagsManifest.items[tag] = data;
+		}
+
+		await this.saveTagsManifest();
+	}
+}
+
+/** Suspense Cache interface for the Cache API. */
+class CacheApiInterface extends CacheInterface<Cache> {
+	constructor(cache: Cache) {
+		super(cache);
+	}
+
+	public override async put(key: string, value: string, init?: RequestInit) {
+		const response = new Response(value, init);
+		await this.cache.put(this.buildCacheKey(key), response);
+	}
+
+	public override async get(key: string) {
+		const response = await this.cache.match(this.buildCacheKey(key));
+		return response ? response.text() : null;
+	}
+
+	public override async delete(key: string) {
+		await this.cache.delete(this.buildCacheKey(key));
+	}
+
+	/**
+	 * Builds the full cache key for the suspense cache.
+	 *
+	 * @param key Key for the item in the suspense cache.
+	 * @returns The fully-formed cache key for the suspense cache.
+	 */
+	public buildCacheKey(key: string) {
+		return `https://${SUSPENSE_CACHE_URL}/entry/${key}`;
+	}
+}
+
+/** Suspense Cache interface for Workers KV. */
+class KVCacheInterface extends CacheInterface<KVNamespace> {
+	constructor(cache: KVNamespace) {
+		super(cache);
+	}
+
+	public override async put(key: string, value: string) {
+		await this.cache.put(key, value);
+	}
+
+	public override async get(key: string) {
+		return this.cache.get(key);
+	}
+
+	public override async delete(key: string) {
+		await this.cache.delete(key);
+	}
+}
+
+/**
+ * Suspense Cache interface for D1.
+ *
+ * **Table Creation SQL**
+ * ```sql
+ * CREATE TABLE IF NOT EXISTS suspense_cache (key text PRIMARY KEY, value text NOT NULL);
+ * ```
+ */
+class D1CacheInterface extends CacheInterface<D1Database> {
+	constructor(cache: D1Database) {
+		super(cache);
+	}
+
+	public override async put(key: string, value: string) {
+		const status = await this.cache
+			.prepare(
+				`INSERT OR REPLACE INTO suspense_cache (key, value) VALUES (?, ?)`,
+			)
+			.bind(key, value)
+			.run();
+		if (status.error) throw new Error(status.error);
+	}
+
+	public override async get(key: string) {
+		const value = await this.cache
+			.prepare(`SELECT value FROM suspense_cache WHERE key = ?`)
+			.bind(key)
+			.first('value');
+		return typeof value === 'string' ? value : null;
+	}
+
+	public override async delete(key: string) {
+		await this.cache
+			.prepare(`DELETE FROM suspense_cache WHERE key = ?`)
+			.bind(key)
+			.run();
+	}
+}
+
+/** Suspense Cache interface for R2. */
+class R2CacheInterface extends CacheInterface<R2Bucket> {
+	constructor(cache: R2Bucket) {
+		super(cache);
+	}
+
+	public override async put(key: string, value: string) {
+		await this.cache.put(key, value);
+	}
+
+	public override async get(key: string) {
+		const value = await this.cache.get(key);
+		return value ? value.text() : null;
+	}
+
+	public override async delete(key: string) {
+		await this.cache.delete(key);
+	}
+}
+
+// TODO: DO Interface
```

Comment:
?

Reply:
For the future, adding a cache interface for Durable Objects. Can remove the TODO if you want, just thought it would be nice to leave in case someone wants to add a Durable Objects one.